OF-353: LDAP

guusdk opened the pull request:
Various LDAP improvements, summed up by:

- Replace usage of String with more appropriate types to represent LDAP data.
- Improve escaping of data.
- Stop wrapping DNs in double-quotes.

More details can be found in the individual commits.

Note: I'm not sure if each individual commit leaves the code in a state that can be compiled without error.

GregDThomas commented on xmppserver/src/test/java/org/jivesoftware/util/LDAPTest.java:
```diff
@@ -25,23 +25,23 @@ public static void reconfigureOpenfireHome() throws Exception {
+//        after = "ou=\"jive\",dc=\"test\",dc=\"jive\",dc=\"com\"";
+//        converted = LdapManager.getEnclosedDN(before);
+//        assertTrue("Conversion result "+before+" to "+converted, converted.equals(after));
+//    }
```
Should this test simply be deleted?

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/ldap/LdapManager.java:
```diff
@@ -1914,20 +1882,19 @@ public void setGroupSearchFilter(String groupSearchFilter) {
+     * @Deprecated As of Openfire 4.5.0, DNs used by Openfire are never enclosed.
+     */
+    @Deprecated
     public void setIsEnclosingDNs(boolean enable) {
```
+1 for keeping old APIs for backward compatability. 

But, enough public APIs have changed (returning LdapName instead of String, taking an LdapName instead of a String parameter) that backward compatability is lost anywau.

Either
a) Add new methods, keeping the old ones, or
b) Remove this old method too.

GregDThomas commented:
Couple of comments, none of any real importance - looks OK to me.

The good news is this looks like something I started a while back, but never finished, that goes towards removing the dependency on that "com.sun" class which will no doubt disappear sooner or later.

speedy01 commented:
@guusdk  got to test this tonight.  objects with forward slashes didn't import.
2019.11.05 20:53:20 ERROR [Jetty-QTP-AdminConsole-146]: org.jivesoftware.openfire.ldap.LdapGroupProvider - Invalid name: "CN=test/test,OU=Security Group - Global"
javax.naming.InvalidNameException: Invalid name: "CN=test/test,OU=Security Group - Global"

another side effect: the ldap admin dn is now strictly requires the full dn.  in the past, this could be userprincipalname format (user.name@domain.com) could be used, but not anymore.  this will likely break some installs for folks (as it did for me! :) )
Got Cache LDAP UserDN was full, shrunk to 90% in 0m multiple times on 1200 users...I don't recall that being an issues in the past, so I will check tomorrow too.  this prob nothing, and I just need to increase the cache size.

guusdk commented:
Thanks @speedy01 - these are very helpful!

Can you please include the first few lines of the stacktrace that I expect was printed in the logs, right after:

```
2019.11.05 20:53:20 ERROR [Jetty-QTP-AdminConsole-146]: org.jivesoftware.openfire.ldap.LdapGroupProvider - Invalid name: "CN=test/test,OU=Security Group - Global" 
javax.naming.InvalidNameException: Invalid name: "CN=test/test,OU=Security Group - Global"
```

> another side effect: the ldap admin dn now strictly requires the full dn. in the past, this could be userprincipalname format (user.name@domain.com) could be used, but not anymore. this will likely break some installs for folks (as it did for me! :) )

I wasn't aware of this. I agree that this is a breaking change. I'll look into it.

> Got Cache LDAP UserDN was full, shrunk to 90% in 0m multiple times on 1200 users

I'm not sure if this is an issue, or if this is us getting good mileage out of our cache. Also, it might be a result of you having tested the shared group issue before?

Fishbowler commented on xmppserver/src/test/java/org/jivesoftware/util/LDAPTest.java:
```diff
@@ -70,32 +70,32 @@ public void testRdnEscapeValue() {
-        String converted = LdapManager.sanitizeSearchFilter(before);
-        assertTrue("Conversion result "+before+" to "+converted, converted.equals(after));
+        String converted = LdapManager.sanitizeSearchFilter(before, false);
+        //assertTrue("Conversion result "+before+" to "+converted+ " expected " + after, converted.equals(after));
```
This has been disabled presumably because it doesn't work.

I've run it and it's now expecting `Test.User\3b \28\2b1\29` - this makes sense given the additional character substitutions in `sanitizeSearchFilter` (https://github.com/igniterealtime/Openfire/pull/1518/files#diff-ff6fe4431447db7be08d7d37b5e4cd02R2301)

Does this just need the `after` updating, or is there more to this?

guusdk replied:
I'm unsure how to proceed with this. One one hand, we could simply update the test to match whatever our code produces now, but that doesn't feel right: we should be testing that whatever is produced is a valid LDAP Search query - not that it matches exactly one of the allowable variants.

speedy01 commented:
@guusdk here you go.  the error for the admin/ldap lookup account and the error on the group. let me know if I can help with anything else.

> 2019.11.05 20:38:00 org.jivesoftware.openfire.ldap.LdapManager - Unable to parse LDAPvalue 'user.name@domain.local'.
javax.naming.InvalidNameException: Invalid name: user.name@domain.local
	at javax.naming.ldap.Rfc2253Parser.doParse(Unknown Source) ~[?:1.8.0_202]
	at javax.naming.ldap.Rfc2253Parser.parseDn(Unknown Source) ~[?:1.8.0_202]
	at javax.naming.ldap.LdapName.parse(Unknown Source) ~[?:1.8.0_202]
	at javax.naming.ldap.LdapName.<init>(Unknown Source) ~[?:1.8.0_202]
	at org.jivesoftware.openfire.ldap.LdapManager.parseAsLdapNameOrLog(LdapManager.java:454) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.ldap.LdapManager.<init>(LdapManager.java:354) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.ldap.LdapManager.<clinit>(LdapManager.java:170) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.ldap.LdapUserProvider.<init>(LdapUserProvider.java:64) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_202]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source) [?:1.8.0_202]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source) [?:1.8.0_202]
	at java.lang.reflect.Constructor.newInstance(Unknown Source) [?:1.8.0_202]
	at java.lang.Class.newInstance(Unknown Source) [?:1.8.0_202]
	at org.jivesoftware.openfire.user.UserManager.initProvider(UserManager.java:500) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.user.UserManager.<init>(UserManager.java:146) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.user.UserManager.<init>(UserManager.java:135) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.user.UserManager.<init>(UserManager.java:56) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.user.UserManager$UserManagerContainer.<clinit>(UserManager.java:85) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.user.UserManager.getInstance(UserManager.java:123) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.XMPPServer.getUserManager(XMPPServer.java:1563) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.spi.PresenceManagerImpl.initialize(PresenceManagerImpl.java:529) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.XMPPServer.initModules(XMPPServer.java:846) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.XMPPServer.start(XMPPServer.java:685) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.XMPPServer.<init>(XMPPServer.java:268) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_202]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source) [?:1.8.0_202]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source) [?:1.8.0_202]
	at java.lang.reflect.Constructor.newInstance(Unknown Source) [?:1.8.0_202]
	at java.lang.Class.newInstance(Unknown Source) [?:1.8.0_202]
	at org.jivesoftware.openfire.starter.ServerStarter.start(ServerStarter.java:92) [startup.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.starter.ServerStarter.main(ServerStarter.java:56) [startup.jar:4.5.0-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_202]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_202]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_202]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_202]
	at com.exe4j.runtime.LauncherEngine.launch(LauncherEngine.java:85) [i4jruntime.jar:?]
	at com.exe4j.runtime.WinLauncher$2.run(WinLauncher.java:90) [i4jruntime.jar:?]
2019.11.05 20:53:20 org.jivesoftware.openfire.ldap.LdapGroupProvider - Invalid name: "CN=test/test,OU=Security Group - Global"
javax.naming.InvalidNameException: Invalid name: "CN=test/test,OU=Security Group - Global"
	at javax.naming.ldap.Rfc2253Parser.parseAttrType(Unknown Source) ~[?:1.8.0_202]
	at javax.naming.ldap.Rfc2253Parser.doParse(Unknown Source) ~[?:1.8.0_202]
	at javax.naming.ldap.Rfc2253Parser.parseDn(Unknown Source) ~[?:1.8.0_202]
	at javax.naming.ldap.LdapName.parse(Unknown Source) ~[?:1.8.0_202]
	at javax.naming.ldap.LdapName.<init>(Unknown Source) ~[?:1.8.0_202]
	at org.jivesoftware.openfire.ldap.LdapManager.findGroupDN(LdapManager.java:1203) ~[xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.ldap.LdapManager.findGroupDN(LdapManager.java:1115) ~[xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.ldap.LdapGroupProvider.getGroup(LdapGroupProvider.java:79) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.group.GroupManager.getGroup(GroupManager.java:359) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.group.GroupManager.getGroup(GroupManager.java:327) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.group.GroupCollection$GroupIterator.getNextElement(GroupCollection.java:115) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.jivesoftware.openfire.group.GroupCollection$GroupIterator.hasNext(GroupCollection.java:76) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at java.util.AbstractCollection.toArray(Unknown Source) [?:1.8.0_202]
	at java.util.ArrayList.<init>(Unknown Source) [?:1.8.0_202]
	at org.jivesoftware.openfire.admin.group_002dsummary_jsp._jspService(group_002dsummary_jsp.java:170) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:70) [apache-jsp-8.5.40.jar:8.5.40]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [javax.servlet-api-3.1.0.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:873) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at com.opensymphony.sitemesh.webapp.SiteMeshFilter.obtainContent(SiteMeshFilter.java:129) [sitemesh-2.4.2.jar:?]
	at com.opensymphony.sitemesh.webapp.SiteMeshFilter.doFilter(SiteMeshFilter.java:77) [sitemesh-2.4.2.jar:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.jivesoftware.util.LocaleFilter.doFilter(LocaleFilter.java:73) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.jivesoftware.util.SetCharacterEncodingFilter.doFilter(SetCharacterEncodingFilter.java:49) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.jivesoftware.admin.PluginFilter.doFilter(PluginFilter.java:226) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.jivesoftware.admin.AuthCheckFilter.doFilter(AuthCheckFilter.java:234) [xmppserver-4.5.0-SNAPSHOT.jar:4.5.0-SNAPSHOT]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1602) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [jetty-security-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1700) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480) [jetty-servlet-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1667) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:220) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:152) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.Server.handle(Server.java:505) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267) [jetty-server-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305) [jetty-io-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [jetty-io-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [jetty-io-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) [jetty-util-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) [jetty-util-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) [jetty-util-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126) [jetty-util-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366) [jetty-util-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698) [jetty-util-9.4.18.v20190429.jar:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804) [jetty-util-9.4.18.v20190429.jar:9.4.18.v20190429]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_202]

speedy01 commented:
@guusdk here you go.  let me know if there is anything else I can help with.
https://pastebin.com/5yTCE31B

guusdk commented:
The commits that I've pushed moments ago should fix the problem with forward slashes in DNs (there's more to the PR, but their effect should be limited to that).

guusdk commented:
I've now reverted representation of the `adminDN` back to `String`. I expect that this resolves the issue that @speedy01 reported (`userprincipalname` format not being usable), but I don't know how to test this. Could someone do that please?

guusdk commented:
I've just rebased this to the current tip of the Master branch.

guusdk commented:
@Fishbowler put me on track of an issue with the latest state of this code. The current code now expects to find a (single) RDN for a user, relative to the Base DN. Example:

    baseDN: o=jivesoftware, o=com
    user RDN: uid=mtucker

It is, however, perfectly valid to have more than one RDN, Example:

    uid=mtucker,ou=People

The current code does not allow for this.

guusdk commented:
I've just commited a change that should address the issue of having more than one RDN in a search result. This change was largely based on @Fishbowler's https://github.com/Fishbowler/Openfire/commit/fb12dfc48b53dfadd1f94ee713a1194fb3e50455 with a few modifications:

- I've used another regular expression to split the search result string into RNDs
- I'm using an array, as opposed to an ArrayList (as the size shouldn't change during the lifecycle of the object).

I've given this a test, both by adding some unit tests as well as trying to reconfigure the scenario with OpenLdap. I didn't find  any problems, but I'm unsure if I correctly configured OpenLdap - further testing is highly desirable.

GregDThomas commented:
I hope to get a gap tomorrow afternoon; I’ll see if I can run it up against ActiveDirectory then.

GregDThomas commented:
Houston, we have a problem.  

Clean build against the master branch, setup against a local AD-based LDAP server, enabled debug, cleaned the logs, restarted, login: All is good.

Recompiled against this branch - and get the following when I attempt to login.

```
2019.11.17 14:13:42 DEBUG [Jetty-QTP-AdminConsole-33]: org.jivesoftware.openfire.ldap.LdapManager - Starting LDAP search for username 'gthomas'...
2019.11.17 14:13:42 DEBUG [Jetty-QTP-AdminConsole-33]: org.jivesoftware.openfire.ldap.LdapManager - Exception thrown when searching for userDN based on username 'gthomas'
javax.naming.NamingException: [LDAP: error code 1 - 000020D6: SvcErr: DSID-0310081B, problem 5012 (DIR_ERROR), data 0
 ]
```

Full logs for both are at https://gist.github.com/GregDThomas/01364333f669e0e9c62709cdc16834f2 in case I don't figure out what the difference is.

GregDThomas commented:
Quick update; no obvious difference apart from the quoting - which AIUI was being removed for a Good Reason (TM):
Good:
```
2019.11.17 14:08:33 DEBUG [Jetty-QTP-AdminConsole-34]: org.jivesoftware.openfire.ldap.LdapManager - LdapManager: Trying to find a user's DN based on their username. sAMAccountName: gthomas, Base DN: OU="Test Standard",DC="test",DC="local"...
```

Bad:
```
2019.11.17 14:13:41 DEBUG [Jetty-QTP-AdminConsole-33]: org.jivesoftware.openfire.ldap.LdapManager - Trying to find a user's RDN based on their username: 'gthomas'. Field: 'sAMAccountName', Base DN: 'OU=Test Standard,DC=test,DC=local' ...
```

guusdk commented:
I'm wondering if @GregDThomas' issue is caused by the space-character in the Base DN. I'm having trouble reproducing this exact scenario. I did, however, reproduce something which I think is pretty similar. That allowed me to identify a bug that also involves the usage os space-characters: the baseDN value is used to construct the LDAP connect URL. The baseDN value was used as-is, which caused an issue: a space has special meaning in the LDAP connect URL syntax (to separate multiple URLS). To avoid issues, I've applied URL encoding to the baseDN value.

GregDThomas commented:
Yup, that's working for me now!

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/ldap/LdapManager.java:
```diff
@@ -1294,20 +1294,26 @@ private Boolean lookupExistence(InitialDirContext ctx, LdapName dn, String[] ret
+            for ( String host : hosts )
+            {
+                // Create a correctly-encoded ldap URL for the PROVIDER_URL
+                final URI uri = new URI("ldap", null, host, port, "/" + baseDN.toString(), null, null);
```
Perhaps 

`sslEnabled ? "ldaps" : "ldap"`

?? (That said, I've got sslEnabled set and it's working fine so I'm guessing the schema isn't used in practice, but will save problems if/when it does start getting used).

guusdk replied:
Yeah, I noticed that too. Want to tackle that in a separate issue. @speedy01 might be interested in this too.

GregDThomas commented:
Something else I just spotted; a search for system properties with description "???" yields three with missing properties; two are not as a result of this merge, but I think:

`ldap.rdn.escape.forwardslash`

is.

[Aside: I introduced `provider.group.groupBasedAdminProvider.groupName` in https://issues.igniterealtime.org/browse/OF-1865 - and whilst I changed the name of the property in the code light of review comments I forgot to change the name of the propertyin the properties that describe it]

GregDThomas commented:
Just found a more major error; go to Server -> Server Settings -> Profile Settings (http://localhost:9090/profile-settings.jsp). Click on "Edit".

```
HTTP ERROR 500

Problem accessing /ldap-server.jsp. Reason:

    Server Error

Caused by:

javax.servlet.ServletException: javax.servlet.ServletException: java.lang.NoSuchMethodError: org.jivesoftware.openfire.ldap.LdapManager.getBaseDN()Ljava/lang/String;
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:168)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:505)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.servlet.ServletException: java.lang.NoSuchMethodError: org.jivesoftware.openfire.ldap.LdapManager.getBaseDN()Ljava/lang/String;
	at org.apache.jasper.runtime.PageContextImpl.doHandlePageException(PageContextImpl.java:909)
	at org.apache.jasper.runtime.PageContextImpl.handlePageException(PageContextImpl.java:838)
	at org.jivesoftware.openfire.admin.ldap_002dserver_jsp._jspService(ldap_002dserver_jsp.java:548)
...
```

guusdk commented:
> Something else I just spotted; a search for system properties with description "???" yields three with missing properties; two are not as a result of this merge, but I think:
> 
> `ldap.rdn.escape.forwardslash`
> 
> is.
>

I've added that property as part of this PR, but ended up not needing it. I have now removed it again.

guusdk commented:
> Just found a more major error; go to Server -> Server Settings -> Profile Settings (http://localhost:9090/profile-settings.jsp). Click on "Edit".
> 
> ```
> HTTP ERROR 500
> 
> Problem accessing /ldap-server.jsp. Reason:
> 
>     Server Error
> 
> Caused by:
> 
> javax.servlet.ServletException: javax.servlet.ServletException: java.lang.NoSuchMethodError: org.jivesoftware.openfire.ldap.LdapManager.getBaseDN()Ljava/lang/String;
> 	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:168)
> 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
> 	at org.eclipse.jetty.server.Server.handle(Server.java:505)
> 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
> 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
> 	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
> 	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
> 	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
> 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
> 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
> 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
> 	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
> 	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
> 	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
> 	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
> 	at java.lang.Thread.run(Thread.java:748)
> Caused by: javax.servlet.ServletException: java.lang.NoSuchMethodError: org.jivesoftware.openfire.ldap.LdapManager.getBaseDN()Ljava/lang/String;
> 	at org.apache.jasper.runtime.PageContextImpl.doHandlePageException(PageContextImpl.java:909)
> 	at org.apache.jasper.runtime.PageContextImpl.handlePageException(PageContextImpl.java:838)
> 	at org.jivesoftware.openfire.admin.ldap_002dserver_jsp._jspService(ldap_002dserver_jsp.java:548)
> ...
> ```

I can't reproduce this.

GregDThomas commented:
> I can't reproduce this.

OK, possibly some stale files hanging around. A clean build removed the problem. Sorry for the noise.

[Aside, not entirely sure that "Use SSL" counts as an "Advanced" setting in 2019]

guusdk commented:
> [Aside, not entirely sure that "Use SSL" counts as an "Advanced" setting in 2019]

I think the TLS side of things needs additional work anyway. We'll leave that for another PR. Kind of want this one to be finished :)

guusdk commented:
@Fishbowler and me identified some issues when using leading and/or trailing space characters in the value used in the 'username' field. We've investigated, but found escaping oddities that we've opted to keep out of this PR.

guusdk commented:
In my understanding, there's no known improvements to be made on this PR. Could you guys please update the review status of this PR when/if you're OK with this being merged?

GregDThomas commented:
If you have't already, don't forget to raise tickets for already identified things (I hesistate to call them shortcomings, this puts the code in a lot better place than it was before). 
It  can remember;
* LDAP TLS improvements
* LDAP usernames with leading/trailing spaces.

guusdk commented:
The leading/trailing space issue is here: https://issues.igniterealtime.org/browse/OF-1922
And this one for TLS: https://issues.igniterealtime.org/browse/OF-1920

Fishbowler commented:
I've tested this on a clean 4.5.0 snapshot plus the built JAR, and as an upgrade from 4.4.2.

Tested with a selection of users (see https://github.com/Fishbowler/tricky-ad-users) and the only aberration (users with leading or trailing whitespace in their CN) has been moved to OF-1922. This is a breaking change (i.e. these users worked in 4.4.2) but fixes users with slashes in their CNs, and possibly other cases not covered by testing, as well as improving the health of the AD code generally.

Tested Admin authentication, listing of users in Users/Groups, user authentication via TCP & BOSH.

Fishbowler commented:
Happy for this to be merged 👍
